lls/main.c

```diff
@@ -209,6 +210,54 @@ submit_nd(struct rte_mbuf **pkts, unsigned int num_pkts,
 	return 0;
 }
 
+/*
+ * Match the packet if it fails to be classifed by ACL rules,
+ * if it's an ND packet, then submit it to the LLS block.
+ */
+static int
+match_nd(struct rte_mbuf *pkt, struct gatekeeper_if *iface)
+{
+	int start;
+	uint8_t nexthdr;
+	uint16_t frag_off;
+	int remaining_len = pkt->data_len;
+	struct ether_hdr *eth_hdr =
+		rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+
+	if (eth_hdr->ether_type == ETHER_TYPE_IPv6) {
+		struct ipv6_hdr *ip6hdr = (struct ipv6_hdr *)&eth_hdr[1];
+		struct icmpv6_hdr *nd_hdr;
+
+		if ((memcmp(ip6hdr->dst_addr, &iface->ip6_addr,
+				sizeof(iface->ip6_addr)) != 0) &&
+				(memcmp(ip6hdr->dst_addr, &iface->ll_ip6_addr,
+				sizeof(iface->ll_ip6_addr)) != 0) &&
+				(memcmp(ip6hdr->dst_addr, &iface->ip6_mc_addr,
+				sizeof(iface->ip6_mc_addr)) != 0) &&
+				(memcmp(ip6hdr->dst_addr,
+				&iface->ll_ip6_mc_addr,
+				sizeof(iface->ll_ip6_mc_addr)) != 0)) {
+			return -1;
+		}
+
+		remaining_len -= sizeof(*eth_hdr) - sizeof(*ip6hdr);
```

Comment:
I think this computation for `remaining_len` is off, and I don't think you need to split up the computation of it. You may not even need a separate variable or it.